src/effects/VST3/VST3ParametersWindow.cpp

```diff
@@ -101,87 +265,86 @@ void VST3ParametersWindow::Fill(Steinberg::Vst::IEditController& editController,
-            const auto value = static_cast<Vst::ParamValue>(slider->GetValue()) / static_cast<Vst::ParamValue>(MaxValue);
-            UpdateParameter(id, value);
-            UpdateParameterValueText(valueText, id, value, unitsStr);
-         });
-         sizer->Add(slider, 0, wxEXPAND);
-         sizer->Add(valueText);
-      }
-      //discrete
       else
       {
-         const auto slider = safenew wxSlider (
-            this,
-            wxID_ANY,
-            static_cast<int>(editController.normalizedParamToPlain(parameterInfo.id, currentNormalizedValue)),
-            0,
-            parameterInfo.stepCount
-         );
-         const auto unitsStr = VST3Utils::ToWxString(parameterInfo.units);
-         const auto valueText = safenew wxStaticText(this, wxID_ANY, wxEmptyString);
-         UpdateParameterValueText(valueText, parameterInfo.id, currentNormalizedValue, unitsStr);
-
-         slider->Bind(wxEVT_SLIDER, [this, id = parameterInfo.id, valueText, unitsStr](wxCommandEvent& event){
-            const auto slider = static_cast<wxSlider*>(event.GetEventObject());
-            UpdateParameter(id, mEditController->plainParamToNormalized(id, slider->GetValue()));
-            UpdateParameterValueText(valueText, id, mEditController->getParamNormalized(id), unitsStr);
-         });
-         sizer->Add(slider, 0, wxEXPAND);
-         sizer->Add(valueText);
+         //continuous
+         if(parameterInfo.stepCount == 0)
+         {
+            auto slider = safenew VST3ContinuousParameter(this, wxID_ANY, parameterInfo.id);
+            slider->UpdateValue(editController);
+            sizer->Add(slider, 0, wxEXPAND);
+            slider->Bind(wxEVT_SLIDER, &VST3ParametersWindow::OnParameterValueChanged, this);
+
+            mControls[parameterInfo.id] = slider;
+         }
+         //discrete
+         else
+         {
+            auto slider = safenew VST3DiscreteParameter(this, wxID_ANY, parameterInfo.id, parameterInfo.stepCount);
+            slider->UpdateValue(editController);
+            sizer->Add(slider, 0, wxEXPAND);
+            slider->Bind(wxEVT_SLIDER, &VST3ParametersWindow::OnParameterValueChanged, this);
+            mControls[parameterInfo.id] = slider;
+         }
+         const auto label = safenew VST3ValueText(this, wxID_ANY, parameterInfo.id, VST3Utils::ToWxString(parameterInfo.units));
+         label->UpdateValue(editController);
+         sizer->Add(label);
+         mLabels[parameterInfo.id] = label;
       }
    }
 
    SetSizer(sizer.release());
 }
 
-void VST3ParametersWindow::UpdateParameterValueText(wxStaticText* text, Steinberg::Vst::ParamID id, Steinberg::Vst::ParamValue normalizedValue, const wxString& units)
+void VST3ParametersWindow::ReloadParameters()
+{
+   for(auto& p : mControls)
+      p.second->UpdateValue(*mEditController);
+   for(auto& p : mLabels)
+      p.second->UpdateValue(*mEditController);
+}
+
+void VST3ParametersWindow::UpdateParameter(Steinberg::Vst::ParamID paramId)
```

Comment:
This and the previous functions do similar things (that, to all parameters, and this, to one parameter only).  Should they have more similar names?

Reply:
I don't think so, `UpdateParameter` and `ReloadParmeters` have different semantics: `UpdateParameter` is to be used when individual parameters are changing during regular plugin usage, `ReloadParmeters` is to be used when plugin state is reset and we don't know which